internal/services/elastic/elastic.go

```diff
@@ -0,0 +1,232 @@
+	// EndTime is the end time of the history.
+	EndTime time.Time
+	// DeviceID is the ID of the device.
+	DeviceID string
+	// PrivilegeIDs is the list of privilege IDs that the user has.
+	PrivilegeIDs []int64
+	// Buckets is the number of time intervals to divide the history into.
+	Buckets int
+}
+
+// GetHistory retrieves the history of a device from elastic. The history is divided into buckets and one data point is selected from each bucket.
+// The data points are selected by the first data point in the bucket that has a vehicle field.
+// The result is a list of data points.
+func (s *Service) GetHistory(ctx context.Context, params GetHistoryParams) ([]any, error) {
+	// set default values for the params
+	setDefaultHistoryParams(&params)
+
+	// build the query
+	req := buildHistoryQuery(params)
+
+	// log requests
+	reqb, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal request: %w", err)
+	}
+	s.log.Info().RawJSON("request", reqb).Str("index", s.settings.DeviceDataIndexNameV2).Msg("V2 request")
+
+	// perform the request
+	res, err := s.esClient.Search().Index(s.settings.DeviceDataIndexNameV2).Request(&req).Perform(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to complete request: %w", err)
+	}
+
+	//nolint:errcheck // we don't care about the error closing the body.
+	defer res.Body.Close()
+
+	// check for errors any status code over 299 is an error
+	if res.StatusCode >= http.StatusMultipleChoices {
+		body, _ := io.ReadAll(res.Body) // we ignore the error here because we are already in an error state
+		return nil, fmt.Errorf("got status code %d from Elastic message = %s.", res.StatusCode, string(body))
+	}
+
+	body, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	// depending on the start and end time we may get len(results) <=  buckets + 1
+	// pass the max number of values to getHitsFromHistoryResponse to ensure we only get the desired number of values
+	return getHitsFromHistoryResponse(body, params.Buckets), nil
+}
+
+// setDefaultHistoryParams sets the default values for the history params.
+func setDefaultHistoryParams(params *GetHistoryParams) {
+	if params.Buckets == 0 {
+		params.Buckets = 1000
+	}
+	if params.EndTime.IsZero() {
+		params.EndTime = time.Now()
+	}
+
+	if params.StartTime.IsZero() {
+		params.StartTime = params.EndTime.Add(-time.Hour * 24 * 14) // default to 2 weeks ago
+	}
+}
+
+// buildHistoryQuery builds the query to retrieve the history of a device from elastic.
+func buildHistoryQuery(params GetHistoryParams) search.Request {
+	// convert the start and end times to milliseconds since epoch
+	startArg := strconv.Itoa(int(params.StartTime.UnixMilli()))
+	endArg := strconv.Itoa(int(params.EndTime.UnixMilli()))
+
+	// calculate the interval in milliseconds
+	interval := calculateInterval(params.StartTime, params.EndTime, params.Buckets)
+	intervalArg := strconv.Itoa(interval) + "ms"
```

Comment:
I'm a little worried about how big their types are here. Already a month is too large for a Java `int`.

Reply:
I will add a test case for this and see how it is handled.

Reply:
It worked out?

Reply:
Yeah I added these test cases https://github.com/DIMO-Network/device-data-api/pull/73/files#diff-5f7a333a7bff5d6ec59400aef69140e2ce90178144803dec0def9180d6379095R293-R314
One creates a large interval value, and the other creates a large number of buckets